python/lsst/cp/pipe/ptc.py

```diff
@@ -369,147 +362,218 @@ def runDataRef(self, dataRef, visitPairs):
-                dataset.rawMeans[ampName].append(mu)
-                dataset.rawVars[ampName].append(varDiff)
-                dataset.inputVisitPairs[ampName].append((v1, v2))
+        return pipeBase.Struct(exitStatus=0)
 
-        numberAmps = len(detector.getAmplifiers())
-        numberAduValues = self.config.maxAduForLookupTableLinearizer
-        lookupTableArray = np.zeros((numberAmps, numberAduValues), dtype=np.float32)
+    def fitNonLinearityAndBuildLinearizers(self, dataset, detector, tableArray=None, log=None):
+        """Fit non-linearity function and build linearizer objects.
 
-        # Fit PTC and (non)linearity of signal vs time curve.
-        # Fill up PhotonTransferCurveDataset object.
-        # Fill up array for LUT linearizer.
-        # Produce coefficients for Polynomial ans Squared linearizers.
-        dataset = self.fitPtcAndNonLinearity(dataset, self.config.ptcFitType,
-                                             tableArray=lookupTableArray)
+        Parameters
+        ----------
+        dataset : `lsst.cp.pipe.ptc.PhotonTransferCurveDataset`
+            The dataset containing information such as the means, variances and exposure times.
 
-        if self.config.makePlots:
-            self.plot(dataRef, dataset, ptcFitType=self.config.ptcFitType)
+        detector : `lsst.afw.cameraGeom.Detector`
+            Detector object.
 
-        # Save data, PTC fit, and NL fit dictionaries
-        self.log.info(f"Writing PTC and NL data to {dataRef.getUri(write=True)}")
-        dataRef.put(dataset, datasetType="photonTransferCurveDataset")
+        tableArray : `np.array`, optional
+            Optional. Look-up table array with size rows=nAmps and columns=DN values.
+            It will be modified in-place if supplied.
 
-        butler = dataRef.getButler()
-        self.log.info("Writing linearizers: \n "
-                      "lookup table (linear component of polynomial fit), \n "
-                      "polynomial (coefficients for a polynomial correction), \n "
-                      "and squared linearizer (quadratic coefficient from polynomial)")
+        log : `lsst.log.Log`, optional
+            Logger to handle messages.
 
-        detName = detector.getName()
+        Returns
+        -------
+        dataset: `lsst.cp.pipe.ptc.PhotonTransferCurveDataset`
+            This is the same dataset as the input paramter, however, it has been modified
+            to include information such as the fit vectors and the fit parameters. See
+            the class `PhotonTransferCurveDatase`.
+
+        linArray : `list`
+            List with (linearizer object, linearizer data type) as entries.
+        """
+
+        # Fit NonLinearity
+        dataset = self.fitNonLinearity(dataset, tableArray=tableArray)
+
+        # Produce linearizer
         now = datetime.datetime.utcnow()
         calibDate = now.strftime("%Y-%m-%d")
 
+        linArray = []
```

Comment:
Why not make the linearity type to fit/create a config option?  Since this code no longer needs to run everything from scratch, there's less of an issue of needing to do all the fits at once.  This simplifies linearity creation, making it more similar to the PTC generation/writes.

Reply:
I have implemented this, but we always fit for the polynomial regardless of the type (that's why the code was persisting everything, since two of the linearizers, as they are implemented now, are by-products of tat polynomial fit).  I think the linearizer squared is redundant with the polynomial. On the other hand, I think is good to have the options split for the reasons you list.